socket_programming(shutdown by Ctrl + C)

Kitaken0107 opened the pull request:
要件：ctrl-c を受け取ったら、サーバーはそれ以上のaccept を行わずに、既にaccept しているリクエストはすべて正常に処理したのちに、サーバーが終了する

Kitaken0107 committed:
socket_programming(shutdown by Ctrl + C)

要件：ctrl-c を受け取ったら、サーバーはそれ以上のaccept を行わずに、既にaccept しているリクエストはすべて正常に処理したのちに、サーバーが終了する

inazz commented on server_ctrlc.c:
```diff
@@ -0,0 +1,114 @@
+
+void signal_handler(int sig) {
+    if (sig == SIGINT) {
+        printf("\nSIGINT received. Shutting down...\n");
```
signal handler 内では printf を呼び出せません。
https://man7.org/linux/man-pages/man7/signal-safety.7.html

inazz commented on server_ctrlc.c:
```diff
@@ -0,0 +1,114 @@
+    int addrlen = sizeof(address);
+
+    // Register signal handler
+    signal(SIGINT, signal_handler);
```
sigaction を使うのが推奨されています。
$ man signal

Kitaken0107 replied:
man signalで確認したところ、
sigactionの方がsignal handlerを使うときに安全と書いていたので、
書き換えてみようと思います

inazz commented on server_ctrlc.c:
```diff
@@ -0,0 +1,114 @@
+        }
+
+        pthread_t thread_id;
+        int *client_socket = malloc(sizeof(int));
```
malloc が失敗したら?

Kitaken0107 replied:
mallocが失敗すると、client_socketにnullが入り、
*client_socket = new_socket;でエラーが発生するので、
*client_socketがnullであるときの検出とcloseが必要かと思いました

Kitaken0107 committed:
Add files via upload

コメント反映させていただきました
・signalをsigactionに変更
・mallocがnullの際のエラーハンドリング
・printfをwriteに変更

Kitaken0107 committed:
Add files via upload

コメント反映版です
